framework/sandstone.cpp

```diff
@@ -2828,6 +2829,89 @@ static void warn_deprecated_opt(const char *opt)
             program_invocation_name, opt);
 }
 
+/* Setup of the performance counters we read to get getloadavg() on linux. */
+#ifdef _WIN32
+
+/* Performance counters we are going to read */
+static const char PQL_COUNTER_PATH[] = "\\System\\Processor Queue Length";
+static HCOUNTER pql_counter;
+
+static const char PT_COUNTER_PATH[]  = "\\Processor(_Total)\\% Processor Time";
+static HCOUNTER pt_counter;
+
+
+static constexpr unsigned TOTAL_5MIN_SAMPLES_COUNT = ((5u * 60u) / 5u);
+static constexpr unsigned SAMPLE_INTERVAL_SECONDS = 5u;
+
+static std::atomic<double> loadavg = 0;
+
+static void loadavg_windows_callback(PVOID, BOOLEAN)
+{
+    PDH_FMT_COUNTERVALUE vpql, vpt;
+
+    if (PdhGetFormattedCounterValue((PDH_HCOUNTER)pql_counter, PDH_FMT_DOUBLE, 0, &vpql) != ERROR_SUCCESS) {
+        return;
+    }
+
+    if (PdhGetFormattedCounterValue((PDH_HCOUNTER)pt_counter, PDH_FMT_DOUBLE, 0, &vpt) != ERROR_SUCCESS) {
+        return;
+    }
+
+    // We calculate current average load as average instantaenous cpu load plus amount of
+    // tasks that are ready to run but cannot be scheduled because CPUs are already
+    // running other tasks.
+    //
+    // We divide by 100.0 to get value in range (0.0;1.0) instead of percents.
+    //
+    // We also mutliply by number of cpus to make the metric behave more like the
+    // /proc/loadavg from Linux, so we get value from range (0.0;num_cpus()), where
+    // num_cpus() value means all cores at 100% utilization.
+    const double current_avg_cpu_usage = (vpt.doubleValue * num_cpus() / 100.0);
+    const double current_proc_queue    = vpql.doubleValue;
+    const double current_avg_load      = current_avg_cpu_usage + current_proc_queue;
+
+    double loadavg_ = loadavg.load(std::memory_order::memory_order_consume);
+    loadavg_ = (current_avg_load + ((TOTAL_5MIN_SAMPLES_COUNT - 1) * loadavg_)) / TOTAL_5MIN_SAMPLES_COUNT;
```

Comment:
This equation won't calculate `loadavg` with the same scaling as Linux does. In Linux, when current average load is 1.0, 1 minute `loadavg` is ~0.62 after 60 seconds, and saturates at 1.0 after roughly 250 seconds.

Here, in the same conditions, after 60 seconds `loadavg_` is ~0.18, and saturates on 60 after ~1500 seconds.

```cpp
loadavg_ = loadavg_ + (SAMPLE_INTERVAL_SECONDS / TOTAL_5MIN_SAMPLES_COUNT) * (current_avg_load - loadavg_)
```
should yield results similar to `\proc\loadavg`. Otherwise we'd have to tune threshold values, and no one wants that :)

Reply:
Equation I suggested should yield results that closely match behavior described in these articles:

* [https://www.brendangregg.com/blog/2017-08-08/linux-load-averages.html](https://www.brendangregg.com/blog/2017-08-08/linux-load-averages.html)
* [https://www.fortra.com/resources/guides/unix-load-average-part-1-how-it-works](https://www.fortra.com/resources/guides/unix-load-average-part-1-how-it-works)

![1, 5, 15 min load averages for a single threaded app](https://www.brendangregg.com/blog/images/2017/loadavg.png)

But let's see what the measurements show.

Reply:
![opendcdiag_loadavg_plot](https://user-images.githubusercontent.com/92308002/204856917-8582e7ee-06ed-4436-a6a6-f02cfc0975d5.jpeg)

This is what I am getting with the code from PR. I have written two simple programs one for Linux and one for Windows.
Biggest difference can be seen when running more jobs than there are cores on the machine and Windows scheduler seem to choke right away, where Linux nicely saturates around expected value.

How to read the legend: 'L' is for Linux and 'W' is for Windows. Numbers with slash are `scheduled_jobs/available_cpus`.
For example `L32/24` means we scheduled 32 jobs on 24 cpu machine.

I was doing the experiment on 8 core Windows machine and 24 Linux machine as you can see, so the graphs will be different in terms of values, but trends are pretty similar and the speed and how they saturate. I think It's close enough and this metric can be useful on Windows for checking if cpu is idle or not, even though it behaves horribly when we have more jobs scheduled than we can run at a given time. I can experiment further with some scaling factor for PQL (\\System\\Processor Queue Length) counter so that it doesn't affect the counter that much. It shouldn't be that hard to do, It's the data collecting that takes most of the time honestly. I'll try to update this data tomorrow with the scaling factor for PQL.

Reply:
We definitely need to scale PQL a bit. Anyway, goal here is to ensure that on Windows idle state can be detected correctly, without tuning the thresholds. If current formula provides that (with some tuning to PQL), I'm ok with it.

Reply:
Aight, got some more numbers.
But first I wanted to say that I misunderstood Linux implementation. For some reason I thought they are scaling nr_uninterruptible somehow, but they are not, It was just implementation of exponential moving average. Today I was analyzing it in more detail and came to this realization.


That being said, I checked how much moving average differs from exponential variant on our calculations. Below graph shows how randomly generated avgload data in two variants. We mock a sudden raise in cpu use and then a drop.
Legend: ma - moving average; ema - exponential moving average
![opendcdiag_loadavg_stats2_2](https://user-images.githubusercontent.com/92308002/205119714-ba73bb9a-f236-47c2-a3e3-97ce9a5a30c1.jpeg)

Some more testing on real data on Windows with ma and ema variant. Exponential variant has "We" in marker name and "W" is with old implementation.
![opendcdiag_loadavg_stats2_0](https://user-images.githubusercontent.com/92308002/205120102-93aabcef-64fd-454b-9bf8-245c3233d752.jpeg)

And some zoom-in on the data when we choke on Windows.
We can try bringing it closer to Linux with some scaling on PQL, but it will vary from system to system, depending on the amount of processes in the background. When we run a busy loop on all other cores, they just sit in the queue, starving. I don't think we can think about any reasonable scaling factor based on that. We would have to do some measurements on scale to know what we decided to use, works.
![opendcdiag_loadavg_stats2_1](https://user-images.githubusercontent.com/92308002/205121184-82dc8b8b-865d-47bc-8e06-f11f228317e7.jpeg)

Reply:
For the purpose of detecting idle state this (EWMA) implementation will be sufficient. I was worried that regular running average would not accumulate value as fast as EWMA does, so we could - potentially - run background scan when we're not supposed to.